packages/react/src/hooks/useIonModal.ts

```diff
@@ -8,14 +8,18 @@ import type { ReactComponentOrElement } from '../models/ReactComponentOrElement'
 import type { HookOverlayOptions } from './HookOverlayOptions';
 import { useOverlay } from './useOverlay';
 
-// TODO(FW-2959): types
-
 /**
  * A hook for presenting/dismissing an IonModal component
  * @param component The component that the modal will show. Can be a React Component, a functional component, or a JSX Element
  * @param componentProps The props that will be passed to the component, if required
  * @returns Returns the present and dismiss methods in an array
  */
+export function useIonModal(component: JSX.Element, componentProps?: any): UseIonModalResult;
+export function useIonModal<P extends undefined>(component: React.ComponentClass<P> | React.FC<P>): UseIonModalResult;
+export function useIonModal<P extends Record<string, never>>(
+  component: React.ComponentClass<P> | React.FC<P>
+): UseIonModalResult;
+export function useIonModal<P>(component: React.ComponentClass<P> | React.FC<P>, componentProps: P): UseIonModalResult;
```

Comment:
Curious on this approach vs. something similar to:

```ts
export function useIonModal<Props = any>(component: JSX.Element | React.ComponentClass<Props> | React.FC<Props>, componentProps?: Props): UseIonModalResult {
```

Appears to accomplish the same level of type inference & type checking.

Reply:
The main reason I did the way I did is so that Typescript requires the second parameter if the component has props.

With your snippet, typescript does not complain if you do not pass props to a component that requires props.

Reply:
Perfect, thanks for the quick response 👍 

I'll chat with Liam on Monday more, but the high chance of type errors in an existing project from this change, makes me lean towards a major - even though this is the correct level of strictness for the implementation. 

Fortunately we are in beta.0 right now, so that means a much shorter window to a release.